File: packages/ai/src/mcp/local-catalog.ts
/**
 * Local user-authored MCP service sources (ENG-6108).
 *
 * Loads a single local JSON file of service entries and validates it against
 * the same contract as the bundled catalog. The loader is deliberately narrow:
 * synchronous file read + JSON parse + structural validation, bounded size and
 * entry counts, no execution, no network, and no credential access —
 * credentials live exclusively in the host's credential storage.
 *
 * The loader rejects anything that would let a local file masquerade as
 * trusted: provenance may only claim the `user` source, and ids colliding
 * with bundled catalog entries are refused (no silent override/rebind of
 * built-ins). How local entries interact with `mcpServers` settings is host
 * policy; this module only provides the validated entries.
 */

import * as fs from "node:fs";
import {
	type McpServiceEntry,
	type McpServiceProvenance,
	SERVICE_CATALOG,
	validateMcpServiceEntry,
} from "./catalog.js";

/** Maximum accepted local source file size. */
export const MAX_LOCAL_CATALOG_BYTES = 256 * 1024;
/** Maximum accepted entries per local source file. */
export const MAX_LOCAL_CATALOG_ENTRIES = 50;

export interface LocalCatalogLoadResult {
	/** Validated local entries, frozen, in file order (ids guaranteed unique and non-bundled). */
	entries: readonly McpServiceEntry[];
	/** Absolute path the entries were loaded from (empty when the file does not exist). */
	path: string;
}

const LOCAL_SOURCE_ALLOWED: Record<string, true> = { user: true };

/**
 * Load and validate a local service source file. A missing file is not an
 * error and yields zero entries; every problem with an existing file throws
 * with the file path (and entry index where applicable) in the message.
 *
 * Proposed settings wiring (host-owned): `~/.prime/agent/mcp-services.json`.
 */
export function loadLocalServiceCatalog(filePath: string): LocalCatalogLoadResult {
	if (!fs.existsSync(filePath)) {
		return { entries: [], path: "" };
	}
	const stat = fs.statSync(filePath);
	if (stat.isDirectory()) {
		throw new Error(`Local service source ${filePath} is a directory, expected a JSON file`);
	}
	if (stat.size > MAX_LOCAL_CATALOG_BYTES) {
		throw new Error(
			`Local service source ${filePath} is ${stat.size} bytes; the maximum is ${MAX_LOCAL_CATALOG_BYTES}`,
		);
	}
	const raw = fs.readFileSync(filePath, "utf8");
	let data: unknown;
	try {
		data = JSON.parse(raw);
	} catch (error) {
		throw new Error(`Local service source ${filePath} is not valid JSON: ${(error as Error).message}`);
	}
	if (typeof data !== "object" || data === null || Array.isArray(data)) {
		throw new Error(`Local service source ${filePath} must be an object with a version and an entries array`);
	}
	const record = data as Record<string, unknown>;
	if (record.version !== 1) {
		throw new Error(`Local service source ${filePath} has unsupported version ${String(record.version)}; expected 1`);
	}
	if (!Array.isArray(record.entries)) {
		throw new Error(`Local service source ${filePath} must contain an entries array`);
	}
	if (record.entries.length > MAX_LOCAL_CATALOG_ENTRIES) {
		throw new Error(
			`Local service source ${filePath} has ${record.entries.length} entries; the maximum is ${MAX_LOCAL_CATALOG_ENTRIES}`,
		);
	}
	const entries: McpServiceEntry[] = [];
	const seenLocal = new Set<string>();
	for (let index = 0; index < record.entries.length; index++) {
		const at = `Local service source ${filePath}, entry ${index}`;
		let entry: McpServiceEntry;
		try {
			entry = validateMcpServiceEntry(record.entries[index]);
		} catch (error) {
			throw new Error(`${at}: ${(error as Error).message}`);
		}
		const badTrust = entry.provenance.find((prov) => !LOCAL_SOURCE_ALLOWED[prov.source]);
		if (badTrust) {
			throw new Error(
				`${at} (${entry.server}): local entries may only carry provenance source "user"; found "${badTrust.source}"`,
			);
		}
		if (entry.provenance.length === 0) {
			throw new Error(
				`${at} (${entry.server}): local entries need at least one provenance record with source "user"`,
			);
		}
		if (seenLocal.has(entry.server)) {
			throw new Error(`${at}: duplicate local id "${entry.server}"`);
		}
		const bundled = SERVICE_CATALOG.find((candidate) => candidate.server === entry.server);
		if (bundled) {
			throw new Error(
				`${at}: id "${entry.server}" collides with the bundled catalog entry for "${bundled.label}"; local sources cannot shadow or rebind built-ins — pick another id`,
			);
		}
		seenLocal.add(entry.server);
		Object.freeze(entry.transport);
		Object.freeze(entry.auth);
		Object.freeze(entry.setup);
		Object.freeze(entry.verification);
		Object.freeze(entry.provenance);
		Object.freeze(entry.aliases);
		Object.freeze(entry.oauth);
		Object.freeze(entry);
		entries.push(entry);
	}
	return { entries, path: filePath };
}

/** Convenience helper for authoring: a provenance record local entries may use. */
export function userProvenance(note: string): McpServiceProvenance {
	return { source: "user", note };
}

File: crates/pa-core/src/mcp/local_catalog.rs
//! Local user-authored MCP service sources (`mcp-services.json`, ENG-6108):
//! a bounded read of one local JSON file of service entries validated
//! against the same contract as the catalog. Port of
//! `packages/ai/src/mcp/local-catalog.ts`.
//!
//! The loader is deliberately narrow: a bounded read of a regular file +
//! JSON parse + structural validation — no execution, no network, no
//! credential access. A local file can never masquerade as trusted:
//! provenance may only claim the `user` source, `legacyBuiltin` and
//! `metadata-reviewed` review status cannot be self-asserted, audit-derived
//! `setup.readiness` cannot be self-asserted, and ids colliding with the
//! compiled built-ins are refused (no silent override/rebind of reserved
//! ids). Errors are visible and bounded: they name the file and entry index
//! and never echo raw input values.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use super::catalog_schema::{McpServiceEntry, ProvenanceSource, VerificationStatus};

/// Maximum accepted local source file size.
pub const MAX_LOCAL_CATALOG_BYTES: u64 = 256 * 1024;
/// Maximum accepted entries per local source file.
pub const MAX_LOCAL_CATALOG_ENTRIES: usize = 50;
/// The local read chunk (bounds how far an oversized read can go).
const READ_CHUNK: usize = 64 * 1024;

/// Validated local entries in file order (ids unique and non-reserved).
#[derive(Debug, Clone, Default)]
pub struct LocalCatalogLoadResult {
    pub entries: Vec<McpServiceEntry>,
    /// The path entries were loaded from; `None` when the file is absent.
    pub path: Option<PathBuf>,
}

/// Envelope of a local source file: `version` (exactly 1) and `entries`.
#[derive(Debug, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct LocalCatalogFile {
    version: u8,
    entries: Vec<McpServiceEntry>,
}

/// Load and validate one local service source file. A missing file is not an
/// error and yields zero entries; every problem with an existing file is a
/// returned error naming the path (and entry index where applicable).
///
/// `reserved` maps every id this installation already owns (compiled
/// built-ins plus the remote/bundled catalog) to its label; a local entry
/// colliding with any of them is refused — local sources cannot shadow or
/// rebind reserved ids.
pub fn load_local_service_catalog(
    path: &Path,
    reserved: &HashMap<String, String>,
) -> Result<LocalCatalogLoadResult, String> {
    if !path.exists() {
        return Ok(LocalCatalogLoadResult::default());
    }
    let metadata = std::fs::metadata(path).map_err(|_| {
        // Vanished between the existence check and stat: treat as absent —
        // the caller re-resolves on the next refresh.
        String::new()
    });
    let Ok(metadata) = metadata else {
        return Ok(LocalCatalogLoadResult::default());
    };
    // Only regular files (symlinks to regular files included): FIFOs,
    // sockets and devices are refused so a special file never hangs the read.
    if !metadata.is_file() {
        return Err(format!(
            "Local service source {} is not a regular file; directories, FIFOs and devices are refused",
            path.display()
        ));
    }
    let bytes = read_bounded(path, MAX_LOCAL_CATALOG_BYTES)?;
    let file: LocalCatalogFile = serde_json::from_slice(&bytes).map_err(|error| {
        let (line, column) = (error.line(), error.column());
        format!(
            "Local service source {} is not valid JSON (parse error near line {line} column {column})",
            path.display()
        )
    })?;
    if file.version != 1 {
        return Err(format!(
            "Local service source {} has an unsupported version; expected 1",
            path.display()
        ));
    }
    if file.entries.len() > MAX_LOCAL_CATALOG_ENTRIES {
        return Err(format!(
            "Local service source {} has {} entries; the maximum is {MAX_LOCAL_CATALOG_ENTRIES}",
            path.display(),
            file.entries.len()
        ));
    }
    let mut entries = Vec::with_capacity(file.entries.len());
    let mut seen_local = std::collections::HashSet::new();
    for (index, entry) in file.entries.iter().enumerate() {
        let at = format!("Local service source {}, entry {index}", path.display());
        entry.validate().map_err(|error| format!("{at}: {error}"))?;
        let id = entry.server.clone();
        if entry
            .provenance
            .iter()
            .any(|prov| prov.source != ProvenanceSource::User)
        {
            return Err(format!(
                "{at} ({id}): local entries may only carry provenance source \"user\""
            ));
        }
        if entry.provenance.is_empty() {
            return Err(format!(
                "{at} ({id}): local entries need at least one provenance record with source \"user\""
            ));
        }
        // Review status and legacy-builtin trust cannot be self-asserted.
        if entry.legacy_builtin {
            return Err(format!(
                "{at} ({id}): local entries cannot claim legacyBuiltin; it is reserved for Prime built-ins"
            ));
        }
        if entry.verification.status != VerificationStatus::Unverified {
            return Err(format!(
                "{at} ({id}): local entries cannot claim \"{}\"; local sources are always unverified",
                match entry.verification.status {
                    VerificationStatus::MetadataReviewed => "metadata-reviewed",
                    VerificationStatus::Unverified => "unverified",
                }
            ));
        }
        if entry.setup.readiness.is_some() {
            return Err(format!(
                "{at} ({id}): local entries cannot claim setup.readiness; readiness is a Prime audit assessment"
            ));
        }
        if !seen_local.insert(id.clone()) {
            return Err(format!("{at}: duplicate local id \"{id}\""));
        }
        if let Some(label) = reserved.get(&id) {
            return Err(format!(
                "{at}: id \"{id}\" collides with the bundled catalog entry for \"{label}\"; local sources cannot shadow or rebind built-ins — pick another id"
            ));
        }
        entries.push(entry.clone());
    }
    Ok(LocalCatalogLoadResult {
        entries,
        path: Some(path.to_path_buf()),
    })
}

/// Bounded read: chunked until the running total exceeds `max_bytes`, so at
/// most `max_bytes` + one chunk is ever read before an oversized file is
/// refused from the read itself.
fn read_bounded(path: &Path, max_bytes: u64) -> Result<Vec<u8>, String> {
    use std::io::Read;
    let file = std::fs::File::open(path)
        .map_err(|_| format!("Local service source {} could not be read", path.display()))?;
    let mut reader = std::io::BufReader::new(file);
    let mut chunks: Vec<u8> = Vec::new();
    let mut total: u64 = 0;
    loop {
        let mut chunk = vec![0u8; READ_CHUNK];
        let read = reader
            .read(&mut chunk)
            .map_err(|_| format!("Local service source {} could not be read", path.display()))?;
        if read == 0 {
            break;
        }
        total += read as u64;
        if total > max_bytes {
            return Err(format!(
                "Local service source {} exceeds the maximum of {max_bytes} bytes",
                path.display()
            ));
        }
        chunk.truncate(read);
        chunks.extend_from_slice(&chunk);
    }
    Ok(chunks)
}
